KAFKA-19434: introduced assignThread method to be able to initialize thread specific resources.

Nikita-Shupletsov opened the pull request:
* introduced assignThread method to be able to initialize thread
specific resources.
* moved all thread specific resources initialization to the new method.
* updated the integration test to verify the fix.

Reviewers: Bill Bejeck<bbejeck@apache.org>

Nikita-Shupletsov committed:
KAFKA-19434: moved the metric initialization to task assignment to have the correct thread.

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -281,6 +281,15 @@ public void shouldPassMetrics(final String topologyType, final String groupProto
+        shouldPassMetrics(topology);
+    }
+
+    private void shouldPassMetrics(Topology topology) throws Exception {
```
`topology` needs to be final

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -281,6 +281,15 @@ public void shouldPassMetrics(final String topologyType, final String groupProto
+        INTERCEPTING_CONSUMERS.clear();
+        INTERCEPTING_ADMIN_CLIENTS.clear();
+
+        shouldPassMetrics(topology);
```
Why do we validate the metrics, clear the clients list, then validate again?
I'm not opposed to this I just want to understand what's going on.

Nikita-Shupletsov replied:
it does pretty much that: `Can we add a test that starts KS validates metrics, stops and starts again validating all metrics?`

the new shouldPassMetrics starts KS and validates metrics.
so we call I call it twice to repeat the same action.

I updated the code a tad to avoid clearing the lists.

Nikita-Shupletsov replied:
with that approach I was able to reproduce the issue without the changes and verify that the changes fix the issue.

bbejeck reviewed (commented):
Thanks @Nikita-Shupletsov for the PR, overall this looks good. 
The `topology` parameter needs to be marked as final for the build to run, Kafka Streams has a checkstyle rule for ensuring variables are marked `final`.

Can we add a test that starts KS validates metrics, stops and starts again validating all metrics?

Nikita-Shupletsov committed:
Passed clientInstance to avoid cleaning the consumers.
Marked the parameters final.

bbejeck reviewed (approved):
Thanks @Nikita-Shupletsov LGTM modulo all tests passing

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -251,8 +251,6 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
                     );
 
                     try {
-                        task.initializeIfNeeded();
```
I am not sure if we can remove this call? We actually get a lock on the state directory here, so I am wondering if it's safe to _not_ get the lock?

But even more crucial: We also register the task on the `stateManager` (that we created above), what is necessary to later access the state store (this code is not written yet, because KIP-1035 is not fully implemented), because we initialize (and open) the state store during registration.

Nikita-Shupletsov replied:
I converted the PR to a draft for now.

I pushed a change and would like to get some feedback:
I extracted the metric creation to a separate method as it was recommended, but there is a metric that's published during init(https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java#L131-L135). I am not 100% sure what to do with it.

also, should the method check the state or something of the task?

mjsax replied:
Good call out -- I believe the comment is actually incorrect? I don't think we would _restore_ anything in this call, but only _register_ the store?

Thus, the metric we measure here, is incorrect to begin with?

mjsax reviewed (commented):
Given that KIP-1035 is not fully implemented yet, it's a hard to reason what the right fix would be, and I believe the proposed fix won't work (cf. inline comment).

It might be necessary, to add a new `Task#initializeMetrics()` method and call within `TaskManager#assignStartupTasks` after the task was assigned to the actual thread, ie, after `assignToStreamThread(...)`?

Maybe @nicktelford who works on KIP-1035 could also take a look?

Nikita-Shupletsov committed:
Extracted metric creation to a separate method to separate task initialization from metric initialization.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractDualSchemaRocksDBSegmentedBytesStore.java:
```diff
@@ -243,15 +246,8 @@ public String name() {
-            taskName,
-            metrics
-        );
+        streamsMetrics = ProcessorContextUtils.metricsImpl(stateStoreContext);
```
I don't think we need to add these two member variables? We keep a reference of `stateStoreContext` as `internalProcessorContext`, and should be able to use `internalProcessorContext` inside `initMetricsIfNeeded` to get `taskId` and the `streamsMetrics` object?

Should apply to other classes, too.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractDualSchemaRocksDBSegmentedBytesStore.java:
```diff
@@ -276,6 +272,19 @@ public void init(final StateStoreContext stateStoreContext, final StateStore roo
+        registerMetrics();
+    }
+
+    private void registerMetrics() {
```
Why are we adding this? Seems, we can just put the code of this method directly into `initMetricsIfNeeded` ?

Seems this also applies to other classes.

Nikita-Shupletsov committed:
Renamed the method to better reflect it's purpose.
Moved thread assignment from init to assignThread in CachingKeyValueStore.
Moved restoreSensor creation to assignThread. it didn't work as is, the behavior will not be that much different.

Nikita-Shupletsov committed:
Small refactoring to address comments.

Nikita-Shupletsov committed:
Merge branch 'trunk' into KAFKA-19434

 Conflicts:
	streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java
	streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java

Nikita-Shupletsov committed:
Made assignThread default with an empty implementation.
Removed empty overrides of the method.

Nikita-Shupletsov commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
as initializeIfNeeded is somewhat idempotent, I am thinking should I add a flag or something for the new method as well to make it idempotent as well? 
@bbejeck @mjsax what do you think?

Nikita-Shupletsov commented:
looks like I missed an important moment that we can use sensors in init. so simply separating the init is not going to work

mjsax commented:
I see your points... That's a beast... 

I was digging into the code a little bit more, and we actually need the "partial initialization" only for `StandbyTasks`, but not for `StreamsTasks`, right?

So I am wondering if adding the new `assignThreads()` method to `Task` interface is actually the right thing to do? Should we instead add a `preInitilize()` (or something like this) method only to `StandbyTasks`? And also add a `StandbyTask#completeInitialization()` that we use to set the thread and init the metrics correctly? -- And we might also fix the state transition, and keep the `StandbyTask` in CREATED state (when calling `preInitilize()`) -- it's not correct that the task going into RUNNING right away, as it does now? And only transit to `RUNNING` inside `completeInitialization()`.

We still need to keep `initializeIfNeeded` and it might call `preInitilize()` and `completeInitlize()` internally to avoid code duplication? This would imply to change `StateDirectory` to use `StandbyTask` (not `Task` to handle startup tasks, but I don't need any disadvantage doing it -- contrary: it seems to be provide better type safety in any case, to use `StandbyTasks` as all startup tasks are `StandbyTask, and thus there is no good reason to use `Task` interface), and only call the new `preInitilize()` method?

Would this work?

Btw: your code is still adding `assignThread()` to `StateStore` what we cannot do w/o a KIP (or we piggy-back it on KIP-1035...)

mjsax commented:
> looks like I missed an important moment that we can use sensors in init. so simply separating the init is not going to work

I think, you "just" need to update `TopologyTestDriver` to also call the new `assignThread` method: https://github.com/apache/kafka/blob/trunk/streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java#L514

As we split the initialization into two calls, calling `task.initializeIfNeeded()` is not sufficient any longer, and leave the task partially initialized, and all test using TDD potentially fail. -- But if we actually don't change the `Task` interface, and don't touch `StreamTask` but only `StandbyTask` as proposed above, the issue should go away by itself.

Nikita-Shupletsov commented:
closing in favor of https://github.com/apache/kafka/pull/20749
